### ISSUE TITLE
[question] Can i apply the X.509 Client Certificate Authentication

### ISSUE THREAD

**Issue by tienvv1234** [moscajs/aedes#545]
I want to create an authentication flow, that's similar to the IoT core,
- X.509 certificates authenticate the client and device connections to the broker
https://docs.aws.amazon.com/iot/latest/developerguide/x509-client-certs.html
if yes, how can I implement it

**Comment by robertsLando** [comment]
Could try to use `preConnect` hook https://github.com/moscajs/aedes/blob/master/docs/Aedes.md#handler-preconnect-client-callback

**Comment by tienvv1234** [comment]
Yes, I can, but I don't know how to send certificates from client to broker and how the broker access to certificates

`const mqtt = require('mqtt');
const client = mqtt.connect('mqtt://localhost:8883', {
    // username: 'test',
    // password: 'test',
    key: fs.readFileSync("./private.pem.key"),
    cert: fs.readFileSync("./certificate.pem.crt"),
});`

**Comment by robertsLando** [comment]
I think you could try to give a look also to https://github.com/moscajs/aedes/blob/master/lib/client.js#L77

I think `decodeProtocol` function isn't documented yet but it's used for example in https://github.com/moscajs/aedes-protocol-decoder to decode the protocol. Try to give a look at them @mcollina thoughts?

**Comment by mcollina** [comment]
This is totally possible and every single time I forget on how to do it. I'd recommend to look on how to do it with bare tcp/Node.js and then port it on top of Aedes/MQTT.

**Comment by robertsLando** [comment]
To parse/validate certs you could use [node-forge](https://www.npmjs.com/package/node-forge)
https://stackoverflow.com/questions/48377731/using-node-js-to-verify-a-x509-certificate-with-ca-cert

@mcollina What I don't understand is when does the parse should happen, I mean what I would do is to do it in `aedes.decodeProtocol`, here I would popolate the client with the `key` and `cert` coming from the connection and in `aedes.authenticate` I can validate the parsed certs using `client.cert` and `client.key`

**Comment by robertsLando** [comment]
Something like: 

```js
var aedes = require('aedes')
var { protocolDecoder, validateCerts } = require('aedes-x509-decoder')
var net = require('net')
var port = 1883

var broker = aedes({
	decodeProtocol: function (client, buffer) {
	  var proto = protocolDecoder(client, buffer) // this will add parsed client.key and client.cert from buffer to client
	  return proto
	},
	preConnect: function (client, done) {
	  validateCerts(client.connDetails.key, client.connDetails.cert, done)
	},
	trustProxy: true
})

var server = net.createServer(broker.handle)

server.listen(port, function () {
  console.log('server listening on port', port)
})
```

**Comment by tienvv1234** [comment]
thank you guys for quick response, I will try it

**Comment by robertsLando** [comment]
@tienvv1234 Let us know :) Could be useful to other users too

**Comment by robertsLando** [comment]
@tienvv1234 Pay attention that with latest release 0.43.0 the preconnect signature has changed

**Comment by tienvv1234** [comment]
ok thank you

**Comment by robertsLando** [comment]
Closing this for now

**Comment by getlarge** [comment]
@robertsLando  Maybe we could keep track of that issue as reference for a feature we could add in the server-factory ?

**Comment by robertsLando** [comment]
Ok I will move this there
